corehq/apps/app_manager/helpers/validators.py

```diff
@@ -0,0 +1,266 @@
+        """
+        visited = set()
+        completed = set()
+
+        def cycle_helper(m):
+            if m.id in visited:
+                if m.id in completed:
+                    return False
+                return True
+            visited.add(m.id)
+            parent = modules.get(neighbour_id_fn(m), None)
+            if parent is not None and cycle_helper(parent):
+                return True
+            completed.add(m.id)
+            return False
+        for module in modules.values():
+            if cycle_helper(module):
+                return True
+        return False
+
+    def _child_module_errors(self, modules_dict):
+        module_errors = []
+
+        def _root_module_fn(module):
+            if hasattr(module, 'root_module_id'):
+                return module.root_module_id
+
+        if self._has_dependency_cycle(modules_dict, _root_module_fn):
+            module_errors.append({'type': 'root cycle'})
+
+        module_ids = set([m.unique_id for m in self.app.get_modules()])
+        root_ids = set([_root_module_fn(m) for m in self.app.get_modules() if _root_module_fn(m) is not None])
+        if not root_ids.issubset(module_ids):
+            module_errors.append({'type': 'unknown root'})
+        return module_errors
+
+    def _check_subscription(self):
+
+        def app_uses_usercase(app):
+            return any(m.uses_usercase() for m in app.get_modules())
+
+        errors = []
+        if app_uses_usercase(self.app) and not domain_has_privilege(self.domain, privileges.USER_CASE):
+            errors.append({
+                'type': 'subscription',
+                'message': _('Your application is using User Properties and your current subscription does not '
+                             'support that. You can remove User Properties functionality by opening the User '
+                             'Properties tab in a form that uses it, and clicking "Remove User Properties".'),
+            })
+        return errors
+
+    @time_method()
+    def validate_app(self):
+        errors = []
+
+        for lang in self.app.langs:
+            if not lang:
+                errors.append({'type': 'empty lang'})
+
+        errors.extend(self._check_modules())
+        errors.extend(self._check_forms())
+
+        if any(not module.unique_id for module in self.app.get_modules()):
+            raise ModuleIdMissingException
+        modules_dict = {m.unique_id: m for m in self.app.get_modules()}
+
+        def _parent_select_fn(module):
+            if hasattr(module, 'parent_select') and module.parent_select.active:
+                return module.parent_select.module_id
+
+        if self._has_dependency_cycle(modules_dict, _parent_select_fn):
+            errors.append({'type': 'parent cycle'})
+
+        errors.extend(self._child_module_errors(modules_dict))
+        errors.extend(self._check_subscription())
+
+        if not errors:
```

Comment:
I know you didn't add this, but I'm curious about why this `if` statement exists. Is it just so the user doesn't get bombarded by many errors at once? Seems like an annoying experience to fix all the errors listed, to then be given another list.

Reply:
That's a good point. It turns out this `if` used to only control a call to `create_all_files`, which probably doesn't work at all if there are other build errors. This changed in https://github.com/dimagi/commcare-hq/commit/b39afd445958d6adead169cc8f63a3f7230be3d3 when `ApplicationBase.validate_app` was added (and was tiny). I don't think it's necessary to gate all of `ApplicationBase.validate_app`. Updated in 462adfcc658876b9b76cacdc042ffac4b2ef20e8